# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/toBESkiii/FlyRank_AI_Intership/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

## 1. My lane and why

My provisional lane is **Refresh / Content Opportunity Scoring**.

I selected this lane because a website may contain thousands of pages, making it difficult for an SEO specialist or content editor to review every page manually. This project will investigate whether search-performance, content and engagement signals can help identify which pages should be reviewed first.

The intended output will be a ranked review queue containing priority scores, suggested actions and clear reasons for each recommendation. The system will support human decision-making rather than automatically deciding that a page must be changed.


In [1]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.

import pandas as pd

DATA_URL = (
    "https://raw.githubusercontent.com/"
    "toBESkiii/FlyRank_AI_Intership/main/"
    "data/raw/content_refresh_anonymized.csv"
)

df = pd.read_csv(DATA_URL)

print("Starter dataset loaded successfully.")
print("Dataset shape:", df.shape)


Starter dataset loaded successfully.
Dataset shape: (30000, 44)


## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

## 2. The question: decision, action, cost of a wrong call

### Research question

**Which website pages should an SEO specialist or content editor review first for refresh, expansion, protection, pruning or monitoring, based on available search-performance, content and engagement signals?**

The unit of analysis is one website page.

The decision being improved is which pages should receive limited editorial attention first. The intended output is a ranked review queue showing the priority of each page, a possible action and the reason for the recommendation.

An SEO specialist or content editor could use the output to investigate the highest-priority pages and decide whether they should be refreshed, expanded, protected, consolidated, pruned or monitored.

A false positive could waste time by recommending a page that does not require attention. A false negative could cause the team to overlook an important page that is losing visibility or has a valuable improvement opportunity.

Data or machine learning may help because page performance can depend on several signals working together. However, the purpose of the project is not simply to train a model. Its purpose is to improve how pages are prioritised for human review.


In [2]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.

# Select pages currently labelled as declining
declining_pages = df[df["trend_direction"] == "down"]

# From those pages, keep pages with at least 100 impressions
declining_with_visibility = declining_pages[
    declining_pages["impressions_90d"] >= 100
]

print("Pages labelled as declining:", len(declining_pages))
print(
    "Declining pages with at least 100 impressions:",
    len(declining_with_visibility)
)


Pages labelled as declining: 16262
Declining pages with at least 100 impressions: 13152


## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

## 3. Quick look at the data

The starter dataset contains **30,000 website pages**.

Of these pages, **16,262 are labelled as declining**, representing approximately **54.2% of the dataset**. In addition, **13,152 pages are both declining and have received at least 100 impressions during the 90-day period**.

These figures show that the dataset contains thousands of possible review candidates. It would be difficult and time-consuming for an SEO specialist or content editor to inspect every declining page manually.

This supports the Refresh / Content Opportunity Scoring lane because a ranked review queue could help the team focus first on pages with stronger evidence of a meaningful opportunity. However, these numbers do not prove that every declining page needs to be refreshed.


In [3]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
# Calculate the total number of pages
total_pages = len(df)

# Count the pages labelled as declining
declining_count = len(declining_pages)

# Count declining pages with at least 100 impressions
visible_declining_count = len(declining_with_visibility)

# Calculate percentages
declining_percentage = declining_count / total_pages * 100
visible_declining_percentage = visible_declining_count / total_pages * 100

print(f"Total pages: {total_pages:,}")
print(
    f"Pages labelled as declining: "
    f"{declining_count:,} ({declining_percentage:.1f}%)"
)
print(
    f"Declining pages with at least 100 impressions: "
    f"{visible_declining_count:,} "
    f"({visible_declining_percentage:.1f}%)"
)

Total pages: 30,000
Pages labelled as declining: 16,262 (54.2%)
Declining pages with at least 100 impressions: 13,152 (43.8%)


## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

## 4. Careful words: what I can and cannot claim

This project will provide **observed, directional and decision-support evidence**.

It may show that certain combinations of search-performance, content and engagement signals are associated with pages receiving higher review-priority scores. It may also show that one prioritisation method performs better than a simple rule-based baseline within the available dataset.

The output may help an SEO specialist or content editor decide which pages deserve investigation first. However, a high-priority score will not automatically mean that a page must be changed.

This project will not claim that it has discovered or can predict Google’s ranking algorithm. It will not prove that a particular feature caused a page to grow or decline. It will also not guarantee that refreshing, expanding or changing a recommended page will improve its future search performance.

A page appearing high in the ranked review queue will only mean that the available data provides stronger reasons to investigate that page before lower-ranked pages.


In [6]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.

# Check whether the decline label matches the trend direction
# Show the columns connected to trends or decline
trend_columns = [
    column for column in df.columns
    if "trend" in column.lower() or "declin" in column.lower()
]

print("Available trend-related columns:")
print(trend_columns)

print("\nNumber of pages in each observed trend category:")
print(df["trend_direction"].value_counts(dropna=False))

print(
    "\nCareful interpretation: these categories describe patterns "
    "observed in the dataset. They do not prove what caused a page "
    "to grow or decline."
)

Available trend-related columns:
['trend_direction', 'trend_pct']

Number of pages in each observed trend category:
trend_direction
down      16262
stable     5962
up         4388
new        2236
flat       1152
Name: count, dtype: int64

Careful interpretation: these categories describe patterns observed in the dataset. They do not prove what caused a page to grow or decline.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.